.dockerignore

```diff
@@ -1,4 +1,3 @@
-.git
```

Comment:
> This would also mean we'd invalidate the cache on any change in the .git directory (which for me happens a lot, e.g. add / remove branches)

Yes indeed cache would be invalidated quite often to load the context here: https://github.com/moby/moby/pull/43529/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R656. I have checked on ci (fetch depth 0) and loading the context with `.git` + building dockerd and docker-proxy takes ~30sec: https://github.com/moby/moby/runs/7425238835?check_suite_focus=true#step:6:2230. Locally it's even faster for me.

> Besides potential metadata (which commit did we build from for inclusion in docker version), are there other reasons we would need this?

Yes this is just to invoke git in https://github.com/moby/moby/pull/43529/files#diff-9eace97aaee3a26c826f37605437498f930565a99a4f072cc476bfeac6199e73R28-R36 if `VERSION`, `GITCOMMIT` or `BUILDTIME` is empty.

> ^^ as, in general, the build should not depend on git (i.e., it should be possible to build from a source tarball, which could be downloaded from the releases page)

Yes we could only rely on ARGs to set `VERSION`, `GITCOMMIT` and `BUILDTIME` for sandboxed builds.

Reply:
> Yes this is just to invoke git in https://github.com/moby/moby/pull/43529/files#diff-9eace97aaee3a26c826f37605437498f930565a99a4f072cc476bfeac6199e73R28-R36 if VERSION, GITCOMMIT or BUILDTIME is empty.

(let me copy that code, as the link will break on rebase)

```bash
if [ -z "$VERSION" ]; then
 	VERSION=$(git describe --match 'v[0-9]*' --dirty='.m' --always --tags | sed 's/^v//' 2> /dev/null || echo "unknown-version")
 fi
 if [ -z "$GITCOMMIT" ]; then
 	GITCOMMIT=$(git rev-parse --short HEAD 2> /dev/null || true)
 fi
 if [ -z "$BUILDTIME" ]; then
 	BUILDTIME=$(date -u --date="@${SOURCE_DATE_EPOCH:-$(date +%s)}" +"%Y-%m-%dT%H:%M:%SZ")
 fi
```

> Yes we could only rely on ARGs to set VERSION, GITCOMMIT and BUILDTIME for sandboxed builds.

Yeah, leaning toward that, as;

- we could already (and already do?) pass that information as argument
- 30 seconds is still 30 seconds 😅 
- for local builds, I don't think the information is really relevant (`dev` as "version" is fine, and I know I just built a binary, so I don't really care about build-time)


Of course happy to hear other opinions

Reply:
`COPY --ignore` would be real nice :)

I think for local dev we'd typically have the source mounted in anyway, in which case we skip the `COPY` step anyway.